compiler/noirc_evaluator/src/acir/tests/instructions/binary.rs

```diff
@@ -0,0 +1,631 @@
+    BLACKBOX::RANGE input: w4, bits: 8
+    BLACKBOX::RANGE input: w5, bits: 8
+    ASSERT w6 = w1 - w5 - 1
+    BLACKBOX::RANGE input: w6, bits: 8
+    ASSERT w5 = -w1*w4 + w0
+    ASSERT w4 = w2
+
+    unconstrained func 0: directive_invert
+    0: @21 = const u32 1
+    1: @20 = const u32 0
+    2: @0 = calldata copy [@20; @21]
+    3: @2 = const field 0
+    4: @3 = field eq @0, @2
+    5: jump if @3 to 8
+    6: @1 = const field 1
+    7: @0 = field field_div @1, @0
+    8: stop &[@20; @21]
+    unconstrained func 1: directive_integer_quotient
+    0: @10 = const u32 2
+    1: @11 = const u32 0
+    2: @0 = calldata copy [@11; @10]
+    3: @2 = field int_div @0, @1
+    4: @1 = field mul @2, @1
+    5: @1 = field sub @0, @1
+    6: @0 = @2
+    7: stop &[@11; @10]
+    ");
+}
+
+#[test]
+fn div_u8_no_predicate_by_constant() {
+    let src = "
+    acir(inline) fn main f0 {
+      b0(v0: u8):
+        v1 = div v0, u8 7
+        return v1
+    }
+    ";
+    let program = ssa_to_acir_program(src);
+
+    // - Note how w2, w3 and w4 are range-checked with less bits than 8
+    // - with `w4 = w3 + 1` and the next range check we ensure that 0 <= w3 < 8
+    // - `w3 = w0 - 7*w2` can be read as `w0 = 7*w2 + w3` which is the definition of divmod
+    assert_circuit_snapshot!(program, @r"
+    func 0
+    private parameters: [w0]
+    public parameters: []
+    return values: [w1]
+    BLACKBOX::RANGE input: w0, bits: 8
+    BRILLIG CALL func: 0, inputs: [w0, 7], outputs: [w2, w3]
+    BLACKBOX::RANGE input: w2, bits: 6
+    BLACKBOX::RANGE input: w3, bits: 3
+    ASSERT w4 = w3 + 1
+    BLACKBOX::RANGE input: w4, bits: 3
+    ASSERT w3 = w0 - 7*w2
+    ASSERT w2 = w1
+
+    unconstrained func 0: directive_integer_quotient
+    0: @10 = const u32 2
+    1: @11 = const u32 0
+    2: @0 = calldata copy [@11; @10]
+    3: @2 = field int_div @0, @1
+    4: @1 = field mul @2, @1
+    5: @1 = field sub @0, @1
+    6: @0 = @2
+    7: stop &[@11; @10]
+    ");
+}
+
+#[test]
+fn div_u8_with_predicate() {
+    let src = "
+    acir(inline) fn main f0 {
+      b0(v0: u8, v1: u8, v2: u1):
+        enable_side_effects v2
+        v3 = div v0, v1
+        return v3
+    }
+    ";
+    let program = ssa_to_acir_program(src);
+
+    // TODO: explain this result
+    assert_circuit_snapshot!(program, @r"
+    func 0
+    private parameters: [w0, w1, w2]
+    public parameters: []
+    return values: [w3]
+    BLACKBOX::RANGE input: w0, bits: 8
+    BLACKBOX::RANGE input: w1, bits: 8
+    BLACKBOX::RANGE input: w2, bits: 1
+    BRILLIG CALL func: 0, inputs: [w1], outputs: [w4]
+    ASSERT w5 = -w1*w4 + 1
+    ASSERT 0 = w1*w5
+    ASSERT 0 = w2*w5
+    BRILLIG CALL func: 1, predicate: w2, inputs: [w0, w1], outputs: [w6, w7]
+    BLACKBOX::RANGE input: w6, bits: 8
+    BLACKBOX::RANGE input: w7, bits: 8
+    ASSERT w8 = w1 - w2 - w7
+    BLACKBOX::RANGE input: w8, bits: 8
+    ASSERT w9 = w1*w6 + w7
+    ASSERT 0 = w0*w2 - w2*w9
+    ASSERT w6 = w3
```

Comment:
These next three asserts are the definition of integer division with predication (`a * predicate == (b * q + r) * predicate`) and the return witness creation.